File: packages/device-client/ios/RNPingDeviceClientCommon.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation
import PingDeviceClient
import React
import RNPingCore

/// Shared execution logic for the Device Client React Native iOS bridge.
///
/// This class maintains a handle-based registry that maps opaque `String`
/// handles (UUIDs) to live `DeviceClient` instances. The registry pattern
/// allows a single React Native app to own multiple clients simultaneously
/// (one per active session, for example). Each JS `createDeviceClient` call
/// produces exactly one registry entry; `dispose` removes it.
///
/// All public entry points are `@objc static` so they can be called from
/// both the TurboModule (`RNPingDeviceClient.mm`) and the classic bridge
/// (`RNPingDeviceClientClassic.mm`) without duplicating logic.
///
/// Ancillary logic lives in sibling files:
/// - ``DeviceErrorMapper`` — maps native errors to `GenericError`
/// - ``DeviceJsonCodec`` — encodes / decodes device payloads
/// - ``DeviceClientConfigNormalizer`` — tolerates sloppy `serverUrl` / `realm` input
@objcMembers
public class RNPingDeviceClientCommon: NSObject {

  /// Lock that serializes all access to ``registry``.
  ///
  /// React Native may invoke bridge methods on a background serial queue
  /// while `dispose` runs on the main thread, so every read/write of
  /// ``registry`` must be guarded by this lock.
  private static let registryLock = NSLock()

  /// Handle-keyed registry of live `DeviceClient` instances.
  ///
  /// Marked `nonisolated(unsafe)` because Swift 6 strict concurrency
  /// checking cannot see the manual `registryLock` protection. Access is
  /// always serialized through ``registryLock``.
  ///
  /// Kept local to this module (not promoted to `RNPingCore`) because no
  /// other package needs to resolve Device Client handles today. Revisit
  /// if a cross-module lookup use case emerges, similar to the shared
  /// logger registry.
  private nonisolated(unsafe) static var registry: [String: DeviceClient] = [:]

  // MARK: - Public entry points

  /// Creates a new `DeviceClient` and stores it in the handle registry.
  ///
  /// The returned handle is an opaque UUID string that JS must pass to every
  /// subsequent `get`, `update`, `deleteDevice`, and `dispose` call.
  ///
  /// - Parameters:
  ///   - config: Configuration dictionary. Required keys: `serverUrl`, `ssoToken`,
  ///     `cookieName`. Optional: `realm` (defaults to `"root"`).
  ///   - resolver: Promise resolver — receives the handle `String` on success.
  ///   - rejecter: Promise rejecter — receives a `GenericError` on failure.
  @objc
  public static func create(
    _ config: NSDictionary,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock
  ) {
    let handlers = PromiseBridge<NSString>(resolver: resolver, rejecter: rejecter)
    guard
      let serverUrl = config["serverUrl"] as? String, !serverUrl.isEmpty,
      let ssoToken = config["ssoToken"] as? String, !ssoToken.isEmpty,
      let cookieName = config["cookieName"] as? String, !cookieName.isEmpty
    else {
      handlers.reject(
        GenericError(
          type: .argumentError,
          error: DeviceClientErrorCode.missingConfig.rawValue,
          message: "serverUrl, ssoToken, and cookieName are required."
        )
      )
      return
    }
    let normalizedServerUrl = DeviceClientConfigNormalizer.normalizeServerUrl(serverUrl)
    let realm = DeviceClientConfigNormalizer.normalizeRealm(config["realm"] as? String)

    // TODO: route JS `loggerId` → `LogManager.logger` (iOS SDK has no per-client
    //   logger field; same global-singleton pattern as the browser package).
    let deviceConfig = DeviceClientConfig(
      serverUrl: normalizedServerUrl,
      realm: realm,
      cookieName: cookieName,
      ssoToken: ssoToken
    )
    let client = DeviceClient(config: deviceConfig)
    let handleId = UUID().uuidString
    registryLock.lock()
    registry[handleId] = client
    registryLock.unlock()
    handlers.resolve(handleId as NSString)
  }

  /// Retrieves all devices of the given type from the server.
  ///
  /// The result is an `NSDictionary` with a single `"result"` key whose value
  /// is an array of encoded device objects.
  ///
  /// - Parameters:
  ///   - handleId: Opaque handle returned by ``create(_:resolver:rejecter:)``.
  ///   - deviceType: One of `DeviceType.oath`, `DeviceType.push`, `DeviceType.bound`, `DeviceType.profile`, or `DeviceType.webAuthn`.
  ///   - resolver: Promise resolver — receives the result dictionary.
  ///   - rejecter: Promise rejecter — receives a `GenericError` on failure.
  @objc
  public static func get(
    _ handleId: String,
    deviceType: String,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock
  ) {
    let handlers = PromiseBridge<NSDictionary>(resolver: resolver, rejecter: rejecter)
    guard let client = lookup(handleId) else {
      handlers.reject(DeviceErrorMapper.handleNotFoundError())
      return
    }
    Task { @MainActor in
      await performGet(client: client, deviceType: deviceType, handlers: handlers)
    }
  }

  /// Updates a single device on the server.
  ///
  /// The `device` dictionary is decoded into the appropriate typed model
  /// (e.g. `OathDevice`, `PushDevice`) before being forwarded to the native SDK.
  ///
  /// - Parameters:
  ///   - handleId: Opaque handle returned by ``create(_:resolver:rejecter:)``.
  ///   - deviceType: One of `DeviceType.oath`, `DeviceType.push`, `DeviceType.bound`, `DeviceType.profile`, or `DeviceType.webAuthn`.
  ///   - device: JS device payload to decode and send.
  ///   - resolver: Promise resolver — receives the updated device dictionary.
  ///   - rejecter: Promise rejecter — receives a `GenericError` on failure.
  @objc
  public static func update(
    _ handleId: String,
    deviceType: String,
    device: NSDictionary,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock
  ) {
    let handlers = PromiseBridge<NSDictionary>(resolver: resolver, rejecter: rejecter)
    guard let client = lookup(handleId) else {
      handlers.reject(DeviceErrorMapper.handleNotFoundError())
      return
    }
    nonisolated(unsafe) let deviceSnapshot = device
    Task { @MainActor in
      await performUpdate(
        client: client,
        deviceType: deviceType,
        device: deviceSnapshot,
        handlers: handlers
      )
    }
  }

  /// Deletes a single device from the server.
  ///
  /// The `device` dictionary is decoded into the appropriate typed model
  /// before being forwarded to the native SDK for deletion.
  ///
  /// - Parameters:
  ///   - handleId: Opaque handle returned by ``create(_:resolver:rejecter:)``.
  ///   - deviceType: One of `DeviceType.oath`, `DeviceType.push`, `DeviceType.bound`, `DeviceType.profile`, or `DeviceType.webAuthn`.
  ///   - device: JS device payload identifying the device to delete.
  ///   - resolver: Promise resolver — receives the deleted device dictionary.
  ///   - rejecter: Promise rejecter — receives a `GenericError` on failure.
  @objc
  public static func deleteDevice(
    _ handleId: String,
    deviceType: String,
    device: NSDictionary,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock
  ) {
    let handlers = PromiseBridge<NSDictionary>(resolver: resolver, rejecter: rejecter)
    guard let client = lookup(handleId) else {
      handlers.reject(DeviceErrorMapper.handleNotFoundError())
      return
    }
    nonisolated(unsafe) let deviceSnapshot = device
    Task { @MainActor in
      await performDelete(
        client: client,
        deviceType: deviceType,
        device: deviceSnapshot,
        handlers: handlers
      )
    }
  }

  /// Removes a `DeviceClient` from the handle registry, releasing its resources.
  ///
  /// After this call, any subsequent operation using the same handle will
  /// receive a `DEVICE_CLIENT_HANDLE_NOT_FOUND` error.
  ///
  /// - Parameters:
  ///   - handleId: Opaque handle returned by ``create(_:resolver:rejecter:)``.
  ///   - resolver: Promise resolver — receives `NSNull` on success.
  ///   - rejecter: Promise rejecter (unused; dispose never fails).
  @objc
  public static func dispose(
    _ handleId: String,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock
  ) {
    registryLock.lock()
    registry.removeValue(forKey: handleId)
    registryLock.unlock()
    resolver(NSNull())
  }

  // MARK: - Dispatch

  /// Dispatches a `get` call to the correct device-type accessor on `client`.
  ///
  /// Runs on `@MainActor` via the calling `Task`. Resolves or rejects through
  /// `handlers` depending on the native SDK result.
  ///
  /// - Parameters:
  ///   - client: The `DeviceClient` instance looked up from the registry.
  ///   - deviceType: Device type key (`DeviceType.oath`, `DeviceType.push`, `DeviceType.bound`, `DeviceType.profile`, `DeviceType.webAuthn`).
  ///   - handlers: Promise bridge that resolves with an `NSDictionary` or rejects with a `GenericError`.
  private static func performGet(
    client: DeviceClient,
    deviceType: String,
    handlers: PromiseBridge<NSDictionary>
  ) async {
    switch deviceType {
    case DeviceType.oath:
      await performFetch(operation: { await client.oath.get() }, handlers: handlers)
    case DeviceType.push:
      await performFetch(operation: { await client.push.get() }, handlers: handlers)
    case DeviceType.bound:
      await performFetch(operation: { await client.bound.get() }, handlers: handlers)
    case DeviceType.profile:
      await performFetch(operation: { await client.profile.get() }, handlers: handlers)
    case DeviceType.webAuthn:
      await performFetch(operation: { await client.webAuthn.get() }, handlers: handlers)
    default:
      handlers.reject(DeviceErrorMapper.invalidDeviceType(deviceType))
    }
  }

  /// Dispatches an `update` call to the correct device-type accessor on `client`.
  ///
  /// Decodes the JS `device` dictionary into the matching `Device` model, then
  /// forwards the typed value to the native SDK.
  ///
  /// - Parameters:
  ///   - client: The `DeviceClient` instance looked up from the registry.
  ///   - deviceType: Device type key (`DeviceType.oath`, `DeviceType.push`, `DeviceType.bound`, `DeviceType.profile`, `DeviceType.webAuthn`).
  ///   - device: Raw JS dictionary to decode into a typed device model.
  ///   - handlers: Promise bridge that resolves with an `NSDictionary` or rejects with a `GenericError`.
  private static func performUpdate(
    client: DeviceClient,
    deviceType: String,
    device: NSDictionary,
    handlers: PromiseBridge<NSDictionary>
  ) async {
    switch deviceType {
    case DeviceType.oath:
      await performMutation(
        OathDevice.self,
        device: device,
        kind: deviceType,
        handlers: handlers,
        operation: { await client.oath.update($0) }
      )
    case DeviceType.push:
      await performMutation(
        PushDevice.self,
        device: device,
        kind: deviceType,
        handlers: handlers,
        operation: { await client.push.update($0) }
      )
    case DeviceType.bound:
      await performMutation(
        BoundDevice.self,
        device: device,
        kind: deviceType,
        handlers: handlers,
        operation: { await client.bound.update($0) }
      )
    case DeviceType.profile:
      await performMutation(
        ProfileDevice.self,
        device: device,
        kind: deviceType,
        handlers: handlers,
        operation: { await client.profile.update($0) }
      )
    case DeviceType.webAuthn:
      await performMutation(
        WebAuthnDevice.self,
        device: device,
        kind: deviceType,
        handlers: handlers,
        operation: { await client.webAuthn.update($0) }
      )
    default:
      handlers.reject(DeviceErrorMapper.invalidDeviceType(deviceType))
    }
  }

  /// Dispatches a `delete` call to the correct device-type accessor on `client`.
  ///
  /// Decodes the JS `device` dictionary into the matching `Device` model, then
  /// forwards the typed value to the native SDK for deletion.
  ///
  /// - Parameters:
  ///   - client: The `DeviceClient` instance looked up from the registry.
  ///   - deviceType: Device type key (`DeviceType.oath`, `DeviceType.push`, `DeviceType.bound`, `DeviceType.profile`, `DeviceType.webAuthn`).
  ///   - device: Raw JS dictionary to decode into a typed device model.
  ///   - handlers: Promise bridge that resolves with an `NSDictionary` or rejects with a `GenericError`.
  private static func performDelete(
    client: DeviceClient,
    deviceType: String,
    device: NSDictionary,
    handlers: PromiseBridge<NSDictionary>
  ) async {
    switch deviceType {
    case DeviceType.oath:
      await performMutation(
        OathDevice.self,
        device: device,
        kind: deviceType,
        handlers: handlers,
        operation: { await client.oath.delete($0) }
      )
    case DeviceType.push:
      await performMutation(
        PushDevice.self,
        device: device,
        kind: deviceType,
        handlers: handlers,
        operation: { await client.push.delete($0) }
      )
    case DeviceType.bound:
      await performMutation(
        BoundDevice.self,
        device: device,
        kind: deviceType,
        handlers: handlers,
        operation: { await client.bound.delete($0) }
      )
    case DeviceType.profile:
      await performMutation(
        ProfileDevice.self,
        device: device,
        kind: deviceType,
        handlers: handlers,
        operation: { await client.profile.delete($0) }
      )
    case DeviceType.webAuthn:
      await performMutation(
        WebAuthnDevice.self,
        device: device,
        kind: deviceType,
        handlers: handlers,
        operation: { await client.webAuthn.delete($0) }
      )
    default:
      handlers.reject(DeviceErrorMapper.invalidDeviceType(deviceType))
    }
  }

  /// Runs a typed fetch operation and forwards the native result to ``handleGet(result:handlers:)``.
  ///
  /// - Parameters:
  ///   - operation: Async fetch operation for a specific device repository.
  ///   - handlers: Promise bridge to resolve or reject.
  private static func performFetch<T: Device & Encodable>(
    operation: () async -> Result<[T], DeviceError>,
    handlers: PromiseBridge<NSDictionary>
  ) async {
    do {
      let result = await operation()
      try handleGet(result: result, handlers: handlers)
    } catch {
      handlers.reject(DeviceErrorMapper.mapError(error))
    }
  }

  /// Decodes a typed device, runs a mutation operation, and forwards the result to ``handleSingle(result:handlers:)``.
  ///
  /// - Parameters:
  ///   - device: Raw JS dictionary to decode.
  ///   - kind: Device kind discriminator used during decode key remapping.
  ///   - handlers: Promise bridge to resolve or reject.
  ///   - operation: Async mutation operation for the typed device.
  private static func performMutation<T: Device & Encodable & Decodable>(
    _: T.Type,
    device: NSDictionary,
    kind: String,
    handlers: PromiseBridge<NSDictionary>,
    operation: (T) async -> Result<T, DeviceError>
  ) async {
    do {
      let typed: T = try DeviceJsonCodec.decode(device, kind: kind)
      let result = await operation(typed)
      try handleSingle(result: result, handlers: handlers)
    } catch {
      handlers.reject(DeviceErrorMapper.mapError(error))
    }
  }

  // MARK: - Result handling

  /// Processes a `get` result that contains an array of devices.
  ///
  /// Each device is individually encoded through ``DeviceJsonCodec/encode(_:)``
  /// (which normalizes timestamps and strips `urlSuffix`) before being wrapped
  /// in a `{ result: [...] }` payload for JS.
  ///
  /// - Parameters:
  ///   - result: Native SDK result containing an array of devices or a `DeviceError`.
  ///   - handlers: Promise bridge to resolve or reject.
  /// - Throws: Re-throws encoding errors from ``DeviceJsonCodec/encode(_:)``.
  private static func handleGet<T: Device & Encodable>(
    result: Result<[T], DeviceError>,
    handlers: PromiseBridge<NSDictionary>
  ) throws {
    switch result {
    case .success(let devices):
      let encoded = try devices.map { try DeviceJsonCodec.encode($0) }
      let payload: NSDictionary = ["result": encoded]
      handlers.resolve(payload)
    case .failure(let error):
      handlers.reject(DeviceErrorMapper.mapDeviceError(error))
    }
  }

  /// Processes an `update` or `delete` result that contains a single device.
  ///
  /// The device is encoded through ``DeviceJsonCodec/encode(_:)`` and wrapped
  /// in a `{ result: { ... } }` payload for JS.
  ///
  /// - Parameters:
  ///   - result: Native SDK result containing a single device or a `DeviceError`.
  ///   - handlers: Promise bridge to resolve or reject.
  /// - Throws: Re-throws encoding errors from ``DeviceJsonCodec/encode(_:)``.
  private static func handleSingle<T: Device & Encodable>(
    result: Result<T, DeviceError>,
    handlers: PromiseBridge<NSDictionary>
  ) throws {
    switch result {
    case .success(let device):
      let encoded = try DeviceJsonCodec.encode(device)
      let payload: NSDictionary = ["result": encoded]
      handlers.resolve(payload)
    case .failure(let error):
      handlers.reject(DeviceErrorMapper.mapDeviceError(error))
    }
  }

  // MARK: - Registry

  /// Thread-safe lookup of a `DeviceClient` by its handle.
  ///
  /// - Parameter handleId: Opaque handle returned by ``create(_:resolver:rejecter:)``.
  /// - Returns: The `DeviceClient` if present, or `nil` if the handle is unknown.
  private static func lookup(_ handleId: String) -> DeviceClient? {
    registryLock.lock()
    defer { registryLock.unlock() }
    return registry[handleId]
  }
}

File: packages/device-client/android/src/main/java/com/pingidentity/rndeviceclient/RNPingDeviceClientCommon.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
package com.pingidentity.rndeviceclient

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReactApplicationContext
import com.pingidentity.android.ContextProvider
import com.pingidentity.device.client.BoundDevice
import com.pingidentity.device.client.Device
import com.pingidentity.device.client.DeviceClient
import com.pingidentity.device.client.DeviceRepository
import com.pingidentity.device.client.OathDevice
import com.pingidentity.device.client.ProfileDevice
import com.pingidentity.device.client.PushDevice
import com.pingidentity.device.client.WebAuthnDevice
import com.pingidentity.logger.Logger
import com.pingidentity.rncore.CoreRuntime
import com.pingidentity.rncore.error.ErrorType
import com.pingidentity.rncore.error.GenericError
import com.pingidentity.rncore.error.reject
import com.pingidentity.rncore.logger.LoggerHandleContract
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import com.pingidentity.rncore.utils.launchBridge
import java.net.MalformedURLException
import java.net.URI
import java.net.URISyntaxException
import java.net.URL
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Shared implementation for the Device Client React Native Android bridge.
 *
 * Maintains a handle → [DeviceClient] registry so a single app can own
 * multiple clients (one per active session, for example).
 *
 * Ancillary logic lives in sibling files:
 * - [DeviceErrorClassifier] — classifies throwables into [GenericError] rejections
 * - [DeviceJson] — encodes / decodes device payloads
 * - [DeviceClientConfigNormalizer] — tolerates sloppy `serverUrl` / `realm` input
 */
object RNPingDeviceClientCommon {

  /**
   * Coroutine scope for executing Device Client operations asynchronously on the IO dispatcher.
   *
   * Uses [SupervisorJob] so that a failure in one coroutine does not cancel siblings.
   */
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

  /**
   * Thread-safe handle-based registry mapping opaque string handles to live [DeviceClient]
   * instances. JS receives the handle id from [create] and passes it back on every
   * subsequent call so the bridge can look up the correct client. The handle is removed
   * when [dispose] is called.
   *
   * Kept local to this module (not promoted to `rncore`) because no other package needs
   * to resolve Device Client handles today. Revisit if a cross-module lookup use case
   * emerges, similar to the shared logger registry.
   */
  private val registry = ConcurrentHashMap<String, DeviceClient>()

  /**
   * Configure application context required by Ping native SDKs.
   *
   * @param reactContext The React Native application context.
   */
  @JvmStatic
  fun configure(reactContext: ReactApplicationContext) {
    ContextProvider.init(reactContext.applicationContext)
  }

  // region  Native entry points

  /**
   * Create a new [DeviceClient] from the supplied configuration and register it
   * in the handle-based registry.
   *
   * Required config keys: `serverUrl`, `ssoToken`, `cookieName`. Optional:
   * `realm` (defaults to `"root"`).
   *
   * On success the promise resolves with the opaque handle id (UUID string)
   * that JS must pass to all subsequent operations.
   *
   * @param config React Native [ReadableMap] containing the client configuration.
   * @param promise React Native promise resolved with the handle id or rejected on validation error.
   */
  @JvmStatic
  fun create(config: ReadableMap, promise: Promise) {
    // TODO-PARITY: Android uses `ssoTokenString` while iOS uses `ssoToken`.
    val serverUrl = config.getStringOrNull("serverUrl")
    val ssoToken = config.getStringOrNull("ssoToken")
    val cookieName = config.getStringOrNull("cookieName")
    if (serverUrl.isNullOrBlank() || ssoToken.isNullOrBlank() || cookieName.isNullOrBlank()) {
      promise.reject(
        GenericError(
          type = ErrorType.ARGUMENT_ERROR,
          error = DeviceClientErrorCodes.DEVICE_CLIENT_MISSING_CONFIG,
          message = "serverUrl, ssoToken, and cookieName are required.",
        ),
      )
      return
    }
    val normalizedServerUrl = DeviceClientConfigNormalizer.normalizeServerUrl(serverUrl)
    val realm = DeviceClientConfigNormalizer.normalizeRealm(config.getStringOrNull("realm"))
    // Parse via URI first (pure RFC-3986 parsing, no potential network
    // resolution) and then convert to URL for the native SDK's constructor.
    // Using URL(String) directly trips CWE-676 in security scanners.
    val parsedUrl: URL = try {
      URI(normalizedServerUrl).toURL()
    } catch (e: URISyntaxException) {
      promise.reject(
        GenericError(
          type = ErrorType.ARGUMENT_ERROR,
          error = DeviceClientErrorCodes.DEVICE_CLIENT_MISSING_CONFIG,
          message = "Invalid serverUrl: ${e.message}",
        ),
        e,
      )
      return
    } catch (e: MalformedURLException) {
      promise.reject(
        GenericError(
          type = ErrorType.ARGUMENT_ERROR,
          error = DeviceClientErrorCodes.DEVICE_CLIENT_MISSING_CONFIG,
          message = "Invalid serverUrl: ${e.message}",
        ),
        e,
      )
      return
    }

    val resolvedLogger = resolveLoggerFromCore(config.getStringOrNull("loggerId"))
    val client = DeviceClient {
      this.ssoTokenString = ssoToken
      this.serverUrl = parsedUrl
      this.realm = realm
      this.cookieName = cookieName
      if (resolvedLogger != null) {
        this.logger = resolvedLogger
      }
    }
    val handleId = UUID.randomUUID().toString()
    registry[handleId] = client
    promise.resolve(handleId)
  }

  /**
   * Resolve a native [Logger] from the shared core logger registry using the
   * opaque handle id emitted by `@ping-identity/rn-logger`.
   *
   * @param id Logger handle id from JS, or null when no logger was supplied.
   * @return Native logger instance, or null when missing / invalid / unsupported.
   */
  private fun resolveLoggerFromCore(id: String?): Logger? {
    if (id.isNullOrBlank()) return null
    val handle = CoreRuntime.loggerRegistry.resolve(id) as? LoggerHandleContract ?: return null
    return handle.nativeLogger as? Logger
  }

  /**
   * Retrieve all devices of the given type from the server.
   *
   * The operation runs asynchronously on the IO dispatcher. On success the
   * promise resolves with a map containing a `result` array of encoded devices.
   *
   * @param handleId Opaque handle id returned by [create].
   * @param deviceType One of [DeviceType.OATH], [DeviceType.PUSH], [DeviceType.BOUND], [DeviceType.PROFILE], or [DeviceType.WEB_AUTHN].
   * @param promise React Native promise resolved with the device list or rejected on error.
   */
  @JvmStatic
  fun get(handleId: String, deviceType: String, promise: Promise) {
    val client = registry[handleId] ?: run {
      DeviceErrorClassifier.rejectHandleNotFound(promise); return
    }
    scope.launchBridge(promise, DeviceClientErrorCodes.DEVICE_CLIENT_ERROR) {
      try {
        // TODO-PARITY: Android repo properties are suffixed with `Device` (`oathDevice`, `pushDevice`, ...)
        //   while iOS uses bare names (`oath`, `push`, ...). Pick one convention.
        // TODO-PARITY: Android uses `.devices()` while iOS uses `.get()`. Pick one convention.
        val result = when (deviceType) {
          DeviceType.OATH -> client.oathDevice.devices()
          DeviceType.PUSH -> client.pushDevice.devices()
          DeviceType.BOUND -> client.boundDevice.devices()
          DeviceType.PROFILE -> client.profileDevice.devices()
          DeviceType.WEB_AUTHN -> client.webAuthnDevice.devices()
          else -> {
            DeviceErrorClassifier.rejectInvalidType(promise, deviceType); return@launchBridge
          }
        }
        result.fold(
          onSuccess = { list ->
            val payload = Arguments.createMap()
            payload.putArray("result", DeviceJson.encodeDevices(list as List<Device>))
            promise.resolve(payload)
          },
          onFailure = { err -> DeviceErrorClassifier.rejectThrowable(promise, err) },
        )
      } catch (e: CancellationException) {
        throw e
      } catch (t: Throwable) {
        DeviceErrorClassifier.rejectThrowable(promise, t)
      }
    }
  }

  /**
   * Update a device on the server.
   *
   * The JS device payload is decoded into the appropriate native [Device]
   * subclass via [DeviceJson.decodeDevice] and forwarded to the matching
   * [DeviceRepository.update] method.
   *
   * @param handleId Opaque handle id returned by [create].
   * @param deviceType One of [DeviceType.OATH], [DeviceType.PUSH], [DeviceType.BOUND], [DeviceType.PROFILE], or [DeviceType.WEB_AUTHN].
   * @param device React Native [ReadableMap] containing the updated device fields.
   * @param promise React Native promise resolved with the updated device or rejected on error.
   */
  @JvmStatic
  fun update(
    handleId: String,
    deviceType: String,
    device: ReadableMap,
    promise: Promise,
  ) {
    val client = registry[handleId] ?: run {
      DeviceErrorClassifier.rejectHandleNotFound(promise); return
    }
    scope.launchBridge(promise, DeviceClientErrorCodes.DEVICE_CLIENT_ERROR) {
      try {
        val decoded = DeviceJson.decodeDevice(deviceType, device)
        val result = when (deviceType) {
          DeviceType.OATH -> client.oathDevice.updateAs<OathDevice>(decoded)
          DeviceType.PUSH -> client.pushDevice.updateAs<PushDevice>(decoded)
          DeviceType.BOUND -> client.boundDevice.updateAs<BoundDevice>(decoded)
          DeviceType.PROFILE -> client.profileDevice.updateAs<ProfileDevice>(decoded)
          DeviceType.WEB_AUTHN -> client.webAuthnDevice.updateAs<WebAuthnDevice>(decoded)
          else -> {
            DeviceErrorClassifier.rejectInvalidType(promise, deviceType); return@launchBridge
          }
        }
        result.fold(
          onSuccess = { d ->
            val payload = Arguments.createMap()
            payload.putMap("result", DeviceJson.encodeDevice(d as Device))
            promise.resolve(payload)
          },
          onFailure = { err -> DeviceErrorClassifier.rejectThrowable(promise, err) },
        )
      } catch (e: CancellationException) {
        throw e
      } catch (t: Throwable) {
        DeviceErrorClassifier.rejectThrowable(promise, t)
      }
    }
  }

  /**
   * Delete a device from the server.
   *
   * The JS device payload is decoded into the appropriate native [Device]
   * subclass via [DeviceJson.decodeDevice] and forwarded to the matching
   * [DeviceRepository.delete] method.
   *
   * @param handleId Opaque handle id returned by [create].
   * @param deviceType One of [DeviceType.OATH], [DeviceType.PUSH], [DeviceType.BOUND], [DeviceType.PROFILE], or [DeviceType.WEB_AUTHN].
   * @param device React Native [ReadableMap] identifying the device to delete.
   * @param promise React Native promise resolved with the deleted device or rejected on error.
   */
  @JvmStatic
  fun deleteDevice(
    handleId: String,
    deviceType: String,
    device: ReadableMap,
    promise: Promise,
  ) {
    val client = registry[handleId] ?: run {
      DeviceErrorClassifier.rejectHandleNotFound(promise); return
    }
    scope.launchBridge(promise, DeviceClientErrorCodes.DEVICE_CLIENT_ERROR) {
      try {
        val decoded = DeviceJson.decodeDevice(deviceType, device)
        val result = when (deviceType) {
          DeviceType.OATH -> client.oathDevice.deleteAs<OathDevice>(decoded)
          DeviceType.PUSH -> client.pushDevice.deleteAs<PushDevice>(decoded)
          DeviceType.BOUND -> client.boundDevice.deleteAs<BoundDevice>(decoded)
          DeviceType.PROFILE -> client.profileDevice.deleteAs<ProfileDevice>(decoded)
          DeviceType.WEB_AUTHN -> client.webAuthnDevice.deleteAs<WebAuthnDevice>(decoded)
          else -> {
            DeviceErrorClassifier.rejectInvalidType(promise, deviceType); return@launchBridge
          }
        }
        result.fold(
          onSuccess = { d ->
            val payload = Arguments.createMap()
            payload.putMap("result", DeviceJson.encodeDevice(d as Device))
            promise.resolve(payload)
          },
          onFailure = { err -> DeviceErrorClassifier.rejectThrowable(promise, err) },
        )
      } catch (e: CancellationException) {
        throw e
      } catch (t: Throwable) {
        DeviceErrorClassifier.rejectThrowable(promise, t)
      }
    }
  }

  /**
   * Remove a [DeviceClient] from the handle-based registry.
   *
   * After this call the handle is no longer valid and any subsequent operations
   * using it will be rejected with [DeviceClientErrorCodes.DEVICE_CLIENT_HANDLE_NOT_FOUND].
   *
   * @param handleId Opaque handle id returned by [create].
   * @param promise React Native promise resolved with null on success.
   */
  @JvmStatic
  fun dispose(handleId: String, promise: Promise) {
    registry.remove(handleId)
    promise.resolve(null)
  }

  // endregion
}

/**
 * Safely read a string from a [ReadableMap], returning null when the key is
 * absent or the value is explicitly null.
 *
 * @param key The field name to read.
 * @return The string value, or null.
 */
private fun ReadableMap.getStringOrNull(key: String): String? =
  if (hasKey(key) && !isNull(key)) getString(key) else null

/**
 * Type-erased update helper that casts the [device] to [T] before delegating
 * to [DeviceRepository.update].
 *
 * This avoids duplicating the `when`-dispatch in the caller for every device
 * type while keeping the generic [DeviceRepository] contract intact.
 *
 * @param device The device instance, expected to be of type [T].
 * @return The result of the update operation.
 */
private suspend inline fun <reified T> DeviceRepository<T>.updateAs(
  device: Any,
): Result<T> {
  require(device is T)
  return this.update(device)
}

/**
 * Type-erased delete helper that casts the [device] to [T] before delegating
 * to [DeviceRepository.delete].
 *
 * This avoids duplicating the `when`-dispatch in the caller for every device
 * type while keeping the generic [DeviceRepository] contract intact.
 *
 * @param device The device instance, expected to be of type [T].
 * @return The result of the delete operation.
 */
private suspend inline fun <reified T> DeviceRepository<T>.deleteAs(
  device: Any,
): Result<T> {
  require(device is T)
  return this.delete(device)
}
